rust/datafusion/src/logicalplan.rs

```diff
@@ -178,54 +174,43 @@ pub enum Expr {
+    pub fn get_type(&self, schema: &Schema) -> Result<DataType> {
         match self {
-            Expr::Column(n) => schema.field(*n).data_type().clone(),
+            Expr::Column(n) => Ok(schema.field(*n).data_type().clone()),
             Expr::Literal(l) => l.get_datatype(),
-            Expr::Cast { data_type, .. } => data_type.clone(),
-            Expr::ScalarFunction { return_type, .. } => return_type.clone(),
-            Expr::AggregateFunction { return_type, .. } => return_type.clone(),
-            Expr::IsNull(_) => DataType::Boolean,
-            Expr::IsNotNull(_) => DataType::Boolean,
+            Expr::Cast { data_type, .. } => Ok(data_type.clone()),
+            Expr::ScalarFunction { return_type, .. } => Ok(return_type.clone()),
+            Expr::AggregateFunction { return_type, .. } => Ok(return_type.clone()),
+            Expr::IsNull(_) => Ok(DataType::Boolean),
+            Expr::IsNotNull(_) => Ok(DataType::Boolean),
             Expr::BinaryExpr {
                 ref left,
                 ref right,
                 ref op,
-            } => {
-                match op {
-                    Operator::Eq | Operator::NotEq => DataType::Boolean,
-                    Operator::Lt | Operator::LtEq => DataType::Boolean,
-                    Operator::Gt | Operator::GtEq => DataType::Boolean,
-                    Operator::And | Operator::Or => DataType::Boolean,
-                    _ => {
-                        let left_type = left.get_type(schema);
-                        let right_type = right.get_type(schema);
-                        get_supertype(&left_type, &right_type).unwrap_or(DataType::Utf8) //TODO ???
-                    }
+            } => match op {
+                Operator::Eq | Operator::NotEq => Ok(DataType::Boolean),
+                Operator::Lt | Operator::LtEq => Ok(DataType::Boolean),
+                Operator::Gt | Operator::GtEq => Ok(DataType::Boolean),
+                Operator::And | Operator::Or => Ok(DataType::Boolean),
+                _ => {
+                    let left_type = left.get_type(schema)?;
+                    let right_type = right.get_type(schema)?;
+                    utils::get_supertype(&left_type, &right_type)
                 }
-            }
+            },
             Expr::Sort { ref expr, .. } => expr.get_type(schema),
         }
     }
 
-    pub fn cast_to(
-        &self,
-        cast_to_type: &DataType,
-        schema: &Schema,
-    ) -> Result<Expr, String> {
-        let this_type = self.get_type(schema);
+    pub fn cast_to(&self, cast_to_type: &DataType, schema: &Schema) -> Result<Expr> {
+        let this_type = self.get_type(schema)?;
         if this_type == *cast_to_type {
             Ok(self.clone())
-        } else if can_coerce_from(cast_to_type, &this_type) {
+        } else {
             Ok(Expr::Cast {
                 expr: Arc::new(self.clone()),
                 data_type: cast_to_type.clone(),
             })
-        } else {
```

Comment:
Did you remove this check because casts support wider types?

Reply:
Added this back in now

Reply:
I think it was fine to remove, which I like better because the user gets the specific reason why the cast failed, as returned by the `cast()` function. My question was more out of curiousity :)